1-Draft/RFC0000-Chain-Operators.md

```diff
@@ -0,0 +1,565 @@
+```
+
+This RFC proposes adding AND-OR lists to PowerShell, using the same `&&` and `||`.
+Because "AND-OR list" is neither intuitive nor common as a terminology,
+instead the term **pipeline chains** is proposed.
+
+## Motivation
+
+> As a PowerShell user, I can chain commands with `&&` and `||`
+> so that I have an ergonomic syntax to conditionally invoke side-effectful commands.
+
+## User Experience
+
+```powershell
+Write-Output "Hello" && Write-Output "Hello again"
+```
+
+```output
+Hello
+Hello again
+```
+
+```powershell
+Write-Output "Hello" || Write-Output "Hello again"
+```
+
+```output
+Hello
+```
+
+```powershell
+Write-Error "Bad" && Write-Output "Hello again"
+```
+
+```output
+Write-Error "Bad" : Bad
++ CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
++ FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException
+```
+
+```powershell
+Write-Error "Bad" || Write-Output "Hello again"
+```
+
+```output
+Write-Error "Bad" : Bad
++ CategoryInfo          : NotSpecified: (:) [Write-Error], WriteErrorException
++ FullyQualifiedErrorId : Microsoft.PowerShell.Commands.WriteErrorException
+
+Hello again
+```
+
+Using `echo` as a convenient example of a successful command
+and `false` as an example of an unsuccesful command.
+
+```powershell
+echo "Hello" && Write-Output "Hi"
+```
+
+```output
+Hello
+Hi
+```
+
+```powershell
+false && Write-Output "Reached"
+```
+
+```output
+```
+
+```powershell
+false || Write-Output "Reached"
+```
+
+```output
+Reached
+```
+
+```powershell
+false || Write-Output "Command failed" && Write-Output "Backup"
+```
+
+```output
+Command failed
+Backup
+```
+
+Also see: [intended test cases for implementation](https://github.com/PowerShell/PowerShell/blob/f8b899e42c86957a1b58273be0029f40a67bf1b6/test/powershell/Language/Operators/BashControlOperator.Tests.ps1).
+
+## Specification
+
+### Grammar
+
+Pipeline chains will be implemented using the already reserved `&&` and `||` operators.
+These operators will have the following grammar:
+
+```none
+statement:
+    | ... <existing statements except pipeline>   # Other statements
+    | variable_expression "=" statement           # Assignment
+    | pipeline_chain ["&"]                        # Pipeline chains
+
+pipeline_chain:
+    | pipeline
+    | pipeline_chain "&&" [newlines] pipeline
+    | pipeline_chain "||" [newlines] pipeline
+```
+
+#### Optional pipeline chaining
+
+A pipeline chain is a chain of one or more pipelines
+and takes the place of a pipeline in the current PowerShell syntax.
+
+In the degenerate case, a pipeline chain of a single pipeline
+will work exactly as pipelines currently do in PowerShell,
+so there is no breakage to the existing grammar.
+
+#### Line continuation
+
+After a pipeline chain operator,
+any newlines will be skipped in anticipation of the following pipeline.
+
+For example, the following would be a single pipeline chain:
+
+```powershell
+cmd1 &&
+    cmd2 ||
+    cmd3
+```
+

```

Comment:
Please consider adding:

Additionally, pipeline chain operators can be placed at the beginning of a line. Newlines before the pipeline chain operators will be skipped in anticipation of the following pipeline.

For example, the following would be a single pipeline chain:

```powershell
cmd1
    && cmd2
    || cmd3
```

Reply:
As said above, we in the @PowerShell/powershell-committee explicitly don't want this for perf and interactiving/scripting consistency reasons.

Reply:
@joeyaiello From an authoring and readability perspective I personally would like this in and be happy with the performance hit that may come with it.

As I believe many others would as well